### ISSUE TITLE
`jdkProxyObj.getClass().getMethod("xxx")` will throw `java.lang.NoSuchMethodException: jdk.proxy4.$Proxy61.xxx()` in the `native-image`.

### ISSUE THREAD

**Issue by wangliang181230** [oracle/graal#6079]
Jdk proxy object can't get method by the `Method Class.getMethod(methodName)` in the `native-image`.

`jdkProxyObj.getClass().getMethod("xxx")` will throw `java.lang.NoSuchMethodException: jdk.proxy4.$Proxy61.xxx()` in the `native-image`.

# Example project

https://github.com/wangliang181230/example__oracle_graal_issue-6079

# Some code

### TestInterface.java
```java
public interface TestInterface {

    void foo();

    void bar();

}
```

### ProxyReflectionExample.java
```java
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

public class ProxyReflectionExample {

    public static void main(String[] args) throws Exception {
        TestInterface proxyObj = (TestInterface)Proxy.newProxyInstance(
            ProxyReflectionExample.class.getClassLoader(),
            new Class[] { TestInterface.class },
            (proxy, method, args1) -> {
                System.out.println("Method: " + method.getName() + ",     Declaring Class: " + method.getDeclaringClass().getName());
                return null;
            });

        try {
            // It will throw `NoSuchMethodException` in `native-image`.
            Method method = proxyObj.getClass().getMethod("foo");
            System.out.println("Method: " + method.getName() + ",     Declaring Class: " + method.getDeclaringClass().getName());
            method.invoke(proxyObj);

            method = proxyObj.getClass().getMethod("bar");
            System.out.println("Method: " + method.getName() + ",     Declaring Class: " + method.getDeclaringClass().getName());
            method.invoke(proxyObj);
        } catch (Throwable t) {
            t.printStackTrace();
        }

        Thread.sleep(10000);
    }
}
```

### reflect-config.json
```json
[
    {
        "name": "TestInterface",
        "methods": [
            { "name": "foo", "parameterTypes": [] },
            { "name": "bar", "parameterTypes": [] }
        ]
    }
]
```

# Console output in `JVM`
```java
Method: foo,     Declaring Class: jdk.proxy1.$Proxy0
Method: foo,     Declaring Class: TestInterface
Method: bar,     Declaring Class: jdk.proxy1.$Proxy0
Method: bar,     Declaring Class: TestInterface
```

# Error log in `native-image`
```java
java.lang.NoSuchMethodException: jdk.proxy4.$Proxy60.foo()
        at java.base@17.0.6/java.lang.Class.getMethod(DynamicHub.java:2227)
        at ProxyReflectionExample.main(ProxyReflectionExample.java:17)
```

# Environment and versions:

1. OS: `Windows 10`
2. GraalVM: `graalvm-ce-java17-22.3.1 Windows (amd64)`

# How can I solve this problem?

**Comment by SergejIsbrecht** [comment]
Do you expect the Proxy to be registered automatically? If not, you would need to register it manually. Please see https://www.graalvm.org/22.0/reference-manual/native-image/DynamicProxy/ -

Also, please add the GraalVM version, which you used.

**Comment by wangliang181230** [comment]
> Do you expect the Proxy to be registered automatically? If not, you would need to register it manually. Please see https://www.graalvm.org/22.0/reference-manual/native-image/DynamicProxy/ -

My question is not about dynamic proxy, but about the acquisition of dynamic proxy methods.
For example `proxyObj.getClass().getMethod(methodName)` will throw `NoSuchMethodException` in `native-image`.
Of course, I have added the metadata of the interfaces to `reflect-config. json`, but the class of the proxy object is unknown.


> Also, please add the GraalVM version, which you used.

graalvm-ce-java17-22.3.1 Windows (amd64)

**Comment by wangliang181230** [comment]
This will cause the semantic inconsistency of `proxyObject.getClass().getMethod(methodName)` in `JVM` and `native-image`.

**Comment by SergejIsbrecht** [comment]
Could you maybe add the stacktrace and the config files to this issue?

**Comment by wangliang181230** [comment]
> Could you maybe add the stacktrace and the config files to this issue?

I modified the content of issue. PTAL

**Comment by oubidar-Abderrahim** [comment]
Thank you for sharing this, we'll take a look into it shortly

**Comment by vjovanov** [comment]
This is currently not supported in the metadata. We should allow for proxies to be marked with an interface list: 
```
  {
        "proxy": ["TestInterface"],
        "methods": [
            { "name": "foo", "parameterTypes": [] },
            { "name": "bar", "parameterTypes": [] }
        ]
    }
```

Until we introduce that feature, a possible workaround is to use initialization at build time. Rewrite the program in the following way:
```
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

interface TestInterface {

    void foo();

    void bar();

}

public class ProxyReflectionExample {

    static class ProxyReflectionStatics {
        static final TestInterface proxyObj = (TestInterface) Proxy.newProxyInstance(
                        ProxyReflectionExample.class.getClassLoader(),
                        new Class[]{TestInterface.class},
                        (proxy, method, args1) -> {
                            System.out.println("Method: " + method.getName() + ",     Declaring Class: " + method.getDeclaringClass().getName());
                            return null;
                        });

    }

    public static void main(String[] args) throws Exception {
        try {
            // It will throw `NoSuchMethodException` in `native-image`.
            Method method = ProxyReflectionStatics.proxyObj.getClass().getMethod("foo");
            System.out.println("Method: " + method.getName() + ",     Declaring Class: " + method.getDeclaringClass().getName());
            method.invoke(ProxyReflectionStatics.proxyObj);

            method = ProxyReflectionStatics.proxyObj.getClass().getMethod("bar");
            System.out.println("Method: " + method.getName() + ",     Declaring Class: " + method.getDeclaringClass().getName());
            method.invoke(ProxyReflectionStatics.proxyObj);
        } catch (Throwable t) {
            t.printStackTrace();
        }

        Thread.sleep(10000);
    }
}
```
and build it with:
```
native-image ProxyReflectionExample --initialize-at-build-time=ProxyReflectionExample\$ProxyReflectionStatics
```

**Comment by wangliang181230** [comment]
@vjovanov 
However, in many applications, it is not known which interface to proxy.

**Comment by sdeleuze** [comment]
Looks like Spring Framework AOP support is impacted by this bug, see [this comment](https://github.com/spring-projects/spring-framework/issues/30529#issuecomment-1596833019) for more details. Could this use case be supported out of the box?

**Comment by sdeleuze** [comment]
I mean without additonal hints specified via JSON files if possible.

**Comment by sdeleuze** [comment]
Indeed, we should be careful about the footprint.

If we specify metadata explicitly, don't we need all the usual variants (invoke versus introspect, declared versus public versus individual methods)?

Should we make introspect the default with the upcoming new reflection mode?

If we need all the variants, should they be in `proxy-config.json`, or in `reflect-config.json` with a list of proxy interfaces instead of a classname (which in this case is not something known before `native-image` invocation?

Will those new entries be backward compatible with GraalVM 22.3.0 and Jube 2023 releases?

**Comment by loicottet** [comment]
Fixed by #8822
